Tactics/SymContext.lean

```diff
@@ -240,32 +224,82 @@ where
     return decl.userName
 
 
-def default (curr_state_number : Nat) : SymContext :=
-  let s := s!"s{curr_state_number}"
-  {
-    state     := .mkSimple s
-    h_run     := .mkSimple s!"h_run"
-    h_program := .mkSimple s!"h_{s}_program"
-    h_pc      := .mkSimple s!"h_{s}_pc"
-    h_err?    := some <| .mkSimple s!"h_{s}_err"
-    h_sp?     := some <| .mkSimple s!"h_{s}_sp"
-    /-
-      `runSteps`, `pc` and `program` actually require inspection of the context.
-      However, these values are not actually used yet,
-      they are merely kept because they will be useful in the future.
-      We can safely put in bogus values for now.
-      (Or we could do the honest thing and make these `Option`s)
-    -/
-    runSteps  := 9999999999
-    program   := `UNUSED
-    pc        := 0#64
-  }
 
-/-! ## Incrementing the context to the next state -/
+/-! ## Massaging the local context -/
 
-/-- `next_state` generates the name for the next intermediate state -/
-def next_state (c : SymContext) : Name :=
-  .mkSimple s!"{c.state_prefix}{c.curr_state_number + 1}"
+/-- If `h_sp` or `h_err` are missing from the `SymContext`,
+add new goals of the expected types,
+and use these to add `h_sp` and `h_err` to the main goal context -/
+def addGoalsForMissingHypotheses (ctx : SymContext) : TacticM SymContext :=
+  withMainContext do
+    let mut ctx := ctx
+    let mut goal ← getMainGoal
+    let mut newGoals := []
+    let lCtx ← getLCtx
+    let some stateExpr :=
+      (Expr.fvar ·.fvarId) <$> lCtx.findFromUserName? ctx.state
+      | throwError "Could not find '{ctx.state}' in the local context"
+
+    if ctx.h_err?.isNone then
+      let h_err? := Name.mkSimple s!"h_{ctx.state}_run"
+      let newGoal ← mkFreshMVarId
+
+      goal := ← do
+        let goalType := h_err_type stateExpr
+        let newGoalExpr ← mkFreshExprMVarWithId newGoal goalType
+        let goal' ← goal.assert h_err? goalType newGoalExpr
+        let ⟨_, goal'⟩ ← goal'.intro1P
+        return goal'
+
+      newGoals := newGoal :: newGoals
+      ctx := { ctx with h_err? }
+
+    if ctx.h_sp?.isNone then
+      let h_sp? := Name.mkSimple s!"h_{ctx.state}_sp"
+      let newGoal ← mkFreshMVarId
+
+      goal := ← do
+        let h_sp_type := h_sp_type stateExpr
+        let newGoalExpr ← mkFreshExprMVarWithId newGoal h_sp_type
+        let goal' ← goal.assert h_sp? h_sp_type newGoalExpr
+        let ⟨_, goal'⟩ ← goal'.intro1P
+        return goal'
+
+      newGoals := newGoal :: newGoals
+      ctx := { ctx with h_sp? }
+
+    replaceMainGoal (goal :: newGoals)
+    return ctx
+
+/-- change the type (in the local context of the main goal)
+of the hypotheses tracked by the given `SymContext` to be *exactly* of the shape
+described in the relevant docstrings.
+
+That is, (un)fold types which were definitionally, but not syntactically,
+equal to the expected shape. -/
+def changeHypothesisTypes (c : SymContext) : TacticM Unit := withMainContext do
```

Comment:
Perhaps name this to be more indicative of its behavior? Maybe `normalize` or `canonicalize` instead of change?

Reply:
I went with canonicalize!